deep_qa/training/trainer.py

```diff
@@ -399,6 +403,90 @@ def evaluate_model(self, data_files: List[str], max_instances: int=None):
+                    # This is a new model object every time.
+                    model = self._build_model()
+                    # We are using the optimizer directly here, so we don't clutter
+                    # the graph creation by not creating optimizers we don't use.
+                    compile_kwargs = self.__compile_kwargs()
+                    compile_kwargs['optimizer'] = None
+                    model.compile(self.__compile_kwargs())
+                    loss = model.total_loss
+                    tower_models.append(model)
+                    grads = self.optimizer.compute_gradients(loss)
+                    tower_gradients.append(grads)
+                    train_loss += loss
+
+        grads = average_gradients(tower_gradients)
+        train_operation = self.optimizer.apply_gradients(grads, global_step=global_step)
+        train_summary = tensorflow.summary.scalar('train_loss', train_loss/self.num_gpus)
+
+        summary_operations = [train_summary]
+        # any metrics that keras has collected
+        merged_metrics = []
+        if tower_models[0].metrics is not None:
+            # merge the metrics across GPUs
+            for i in range(len(tower_models[0].metrics)):
+                name = tower_models[0].metrics[0]
+                tensor = tensorflow.reduce_mean([mm.metrics_tensors[i] for mm in tower_models])
+                summary_operations.append(tensorflow.summary.scalar(name, tensor))
+                merged_metrics.append(tensor)
+
+        inputs = []
+        updates = []
+        for model in tower_models:
+            # pylint: disable=protected-access
+            model_inputs = (model._feed_inputs + model._feed_targets + model._feed_sample_weights)
+            # pylint: enable=protected-access
+            inputs.extend(model_inputs)
+            updates.extend(model.updates)
+        # Just check any one, as we just made copies of them.
+        if tower_models[0].uses_learning_phase and \
+                not isinstance(K.learning_phase(), int):
+            inputs += [K.learning_phase()]
+
+        if self.tensorboard_log is not None:
+            train_summary_writer = tensorflow.summary.FileWriter(os.path.join(self.tensorboard_log, "train"))
+        else:
+            train_summary_writer = None
+
+        # Add the multi-gpu update operation.
+        updates += [train_operation]
+        # Gets loss and metrics. Updates weights at each call.
+        primary_model = tower_models[0]
+        primary_model.train_function = Step(inputs,
+                                            [train_loss] + merged_metrics,
+                                            global_step,
+                                            summary_writer=train_summary_writer,
+                                            summary_frequency=self.tensorboard_frequency,
+                                            updates=updates)
+        # TODO(Mark): This is a bit hacky. Add to compile instead?
```

Comment:
Oh, I see what you mean.  Instead of checking for `hasattr` in `model._fit_loop`, you just always add an attribute to `DeepQaModel` that will be `1` for the tower models, but `>1` for the primary model?  Seems reasonable.